[I2C] Implement I2C Multiplexer

tonhuisman opened the pull request:
Implementation of generic/global I2C Multiplexer support.
Supports TCA9548A (8 channels, select 1 of 8 I2C addresses, tested) and TCA9546A (untested, is very similar but only supports 4 channels). Prepared for PCA9540 (untested, 2 channels, fixed I2C address, somewhat different control).

- Hardware page has selection input for Multiplexer device and I2C address.
- Device overview page shows I2C Multiplexer channel if configured for that device.
- Device configuration page for I2C device shows Multiplexer channel selection if configured completely in Hardware page.
- PluginCall does the 'heavy lifting' of switching to the correct multiplexer channel, and off again, if configured,
- I2CScanner shows an extra column I2C Bus if a multiplexer is configured, and scans all available channels for devices, including `/i2cscanner_json` when that is enabled.
- `/json` endpoint includes `I2CBus` field if a multiplexer is configured.
- Entire implementation can be disabled by commenting `#define FEATURE_I2CMULTIPLEXER` in `define_plugin_sets.h`. Disabled for 'minimal', 'switch_only' and 'IR' builds
- Updated build to `20110` to properly initialize the new settings.
- This should 'obsolete' PR #254 
- todo: Update documentation & screenshots

NB: Adds just below 3.4 kB to the ESP8266 build and a little more than 4.5 kB to the ESP32 build (.bin size)

Some screenshots: (mashed some I2C devices available onto an ESP32)
*I2C Scanner output:*
![I2CScanner_with_multiplexer](https://user-images.githubusercontent.com/1922393/90176426-2ea50680-dda9-11ea-9a12-748e6951822b.png)

*Hardware page, select I2C Multiplexer:*
![Hardware_page_select_multiplexer_type](https://user-images.githubusercontent.com/1922393/90176507-585e2d80-dda9-11ea-8ef8-7107f7c21653.png)

*Hardware page, select multiplexer address:*
![Hardware_page_select_multiplexer_address](https://user-images.githubusercontent.com/1922393/90176569-762b9280-dda9-11ea-9e76-c525c625ed99.png)

*Devices list, with some multiplexed devices:*
![Devices_list_with_multiplexed_devices](https://user-images.githubusercontent.com/1922393/90176670-9d825f80-dda9-11ea-817e-9f4e1c7ab037.png)

*Device page, with multiplexer enabled, selection of Multiplexer channel:*
![Device_page_select_multiplexer_channel](https://user-images.githubusercontent.com/1922393/90176744-bab72e00-dda9-11ea-92ad-92c5273dcbec.png)

tonhuisman committed:
[I2C] Implement I2C Multiplexer

TD-er commented:
The build errors for ESP32 builds:

```
Compiling .pio/build/custom_ESP32_4M316k/src/src/Helpers/ESPEasy_checks.cpp.o
src/src/Helpers/ESPEasy_checks.cpp: In instantiation of 'void check_size() [with ToCheck = SettingsStruct_tmpl<32u>; unsigned int ExpectedSize = 2900u; unsigned int RealSize = 2936u]':
```

For ESP8266:
```
Compiling .pio/build/custom_ESP8266_4M1M/src/src/Helpers/ESPEasy_checks.cpp.o
src/src/Helpers/ESPEasy_checks.cpp: In instantiation of 'void check_size() [with ToCheck = SettingsStruct_tmpl<12u>; unsigned int ExpectedSize = 1260u; unsigned int RealSize = 1276u]':
```

Quick fix in ESPEasy_checks.cpp:
```c++
  const unsigned int SettingsStructSize = (280 + 83 * TASKS_MAX);
```

TD-er commented on src/WebServer_DevicesPage.ino:
```diff
@@ -704,6 +718,32 @@ void handle_devices_TaskSettingsPage(taskIndex_t taskIndex, byte page)
+      addFormSubHeader(F("I2C Multiplexer"));
+
+      int taskDeviceI2CMuxPort = Settings.I2C_Multiplexer_Port[taskIndex];
+      String  i2c_mux_portoptions[9];
```
In general, when using these stack allocated arrays (N.B. also is applicable for heap allocated), please wrap their definition and usage in a small scope using 
```c++
{
  String bla[9];
   ...
   function(bla);
}
```
This way, the data is removed from the stack as soon as you try to call the other functions.
And since the `String` object is both stack and heap allocated, you may just prevent a crash due to memory issues when these Strings may have become large or fragment the heap.

tonhuisman replied:
Ah, yes, I try to do that every time, but this one I forgot, it seems.

TD-er commented on src/src/DataStructs/SettingsStruct.h:
```diff
@@ -194,6 +194,9 @@ class SettingsStruct_tmpl
   uint8_t       ETH_Wifi_Mode;
+  int8_t        I2C_Multiplexer_Addr;
+  int8_t        I2C_Multiplexer_Port[N_TASKS];
+  int8_t        I2C_Multiplexer_Type;
```
Better have the type before the port values.
This way we have the nr task related options at the end (for as long as it may last)

tonhuisman replied:
The Type field was an afterthought, when I already ran it on both an ESP32 and ESP8266. I'll move it around, and clear/fix my local settings, of course.

TD-er commented on src/define_plugin_sets.h:
```diff
@@ -23,6 +23,7 @@ To create/register a plugin, you have to :
 */
 
 //#define FEATURE_SD
+#define FEATURE_I2CMULTIPLEXER
```
```c++
#ifndef FEATURE_I2CMULTIPLEXER
    #define FEATURE_I2CMULTIPLEXER
#endif
```

If you set the define in the custom Python definitions, you may get lots of warnings about defines being redefined.

TD-er replied:
You undefine it also in the BUILD_ONLY_SWITCH, but the other entries there only 'add' stuff.
So I guess it should be defined in the 'normal' build and undefined when space is needed.

```c++
#ifdef PLUGIN_BUILD_NORMAL
    #define  PLUGIN_SET_STABLE
    #define  CONTROLLER_SET_STABLE
    #define  NOTIFIER_SET_STABLE

    #ifndef FEATURE_I2CMULTIPLEXER
        #define FEATURE_I2CMULTIPLEXER
    #endif

    #ifndef BUILD_NO_DEBUG
      #define BUILD_NO_DEBUG
    #endif
#endif
```

This also allows it to be built using Custom builds.

tonhuisman replied:
I'll fix that part, I now see it should be done somewhat different. Thnx.

tonhuisman replied:
I'm thinking of *only* setting this define from the environment configuration files, as I kind of deemed it as a feature, that needs to be enabled explicitly. And then I'd set it only for the 4 MB builds.

TD-er replied:
Just keep in mind these 2 simple "rules":
- Everything must be configurable from a "custom" build.
- Builds aimed for limited resources generally use blocks to specifically "disable" parts.

If you keep those in mind, I guess it can be rather simple to decide where to put this check.
Just make sure to wrap it in #ifndef like I showed, since it could already be defined explicitly.

tonhuisman replied:
I now only define it in the `PLUGIN_BUILD_NORMAL` configuration, and undef it in the `BUILD_MINIMAL_OTA` or `MEMORY_ANALYSIS` special configuration, so it is effectively included in most normal builds. Probably some that now go over their size need this to be undef'ed, we'll see what builds travis chokes on.

TD-er commented:
I've been thinking about this a bit ( ;) ) more and there is a number of things I realized:

- All multiplexer ports disabled is like "direct connected", but the main I2C bus is always connected.
- It is possible to switch frequency to a lower freq. when trying to communicate to a specific multiplexer channel (#3208 , #3207) Maybe add a checkmark or a combo box to force an output channel to a lower frequency?
- You can set any combination of ports on the multiplexer, so why not make the setting value a bitmask?
- The multiplexer chip has a reset pin, maybe also add support for that via an optional GPIO pin?

The first point may need some explaining.
If anything is not connected via the multiplexer, but directly to the ESP's I2C bus, then it will come up on any scan.
Also it is always present regardless the state of the multiplexer chip.
This also has an effect on the 2nd point I mentioned as you can clock down the frequency without issues, but up may not always work.
The documentation should therefore mention a number of use cases, where "old"/"slow" sensors should be put on a separate channel.

I don't know if it is practical to have multiple outputs active at the same time, but the chip supports it, so I think we should at least let the settings we use reflect how it is being used on the chip. On the other hand, if we later want to support multiple of these, we may quickly run out of "bits" in our settings. What do you think?

TD-er commented:
As far as I remember, I do have a standard function to add the I2C Address field.
Why not add the multiplexer port selector to that same function? Then the I2C settings are always together.
Maybe it should be outside the other settings like you now also try to do with the multiplexer.

For example the INA219 now does look a bit strange:

![image](https://user-images.githubusercontent.com/3751318/90222800-1666d300-de0d-11ea-9d8e-eb6da0a974dd.png)

The same for the BME280:
![image](https://user-images.githubusercontent.com/3751318/90222878-3b5b4600-de0d-11ea-9483-3cdb5b55b712.png)

TD-er commented:
One other thing... 
Support for the I2C UART chip I added (used by ESPEasy serial) is not possible to be used via the multiplexer like it is now implemented in this PR.
At least the _CPlugin code must then also perform the switching if we also need this one to operate behind the multiplexer.

tonhuisman commented:
> Quick fix in ESPEasy_checks.cpp:

Will fix, I would have expected that check to be enabled in the normal and testing builds (that's what I've been developing on), so I first thought the failing builds were caused by size issues.

> * All multiplexer ports disabled is like "direct connected", but the main I2C bus is always connected.

Don't exactly understand what you're trying to say here?

> * It is possible to switch frequency to a lower freq. when trying to communicate to a specific multiplexer channel (#3208 , #3207) Maybe add a checkmark or a combo box to force an output channel to a lower frequency?

That would mean we'd either define `low speed` quantified fixed at 100 kHz, or introduce a new setting for the definition of low speed. Sounds useful.
And then also a bitmapped setting for the channels requiring low speed (think I'll use an `uint8_t` or similar for that, as I don't expect to find (or need) a multiplexer with > 8 channels)

> * You can set any combination of ports on the multiplexer, so why not make the setting value a bitmask?

Thought about that, but I see no real reason for slave I2C devices to be able to see each other on the bus (and you can always place them on the same channel if that would be needed, as then they most likely won't have colliding addresses, the main argument for using a multiplexer, IMHO). And we don't (intend to) support multiple-master I2C, AFAIK?
I could provide a set of checkboxes per the number of channels, and use these as a bitmap to send to the multiplexer, instead of shifting a bit to the desired position. Again the maximum number of channels would be limited to 8. And I haven't seen any I2C multiplexers with over 8 channels yet.

> * The multiplexer chip has a reset pin, maybe also add support for that via an optional GPIO pin?

Well, under what conditions would you programmatically reset the multiplexer? I think it should be tied to the reset of the controller, so when that is pulled down, the multiplexer is also reset.

> as you can clock down the frequency without issues, but up may not always work.

That will involve testing with an actual `DEVICE_TYPE_I2C_LOW_SPEED` (to be implemented) device, and I already found out that the multiplexer itself does not support the 700 kHz speed of the OLED (it is rated at max 400 kHz), as the OLEDs were acting 'funny' when connected behind the multiplexer on the ESP32 (that seems to be capable of that speed), at first (until I removed the initialization from the OLED driver).

> As far as I remember, I do have a standard function to add the I2C Address field.
> Why not add the multiplexer port selector to that same function? Then the I2C settings are always together.

I tried to do that, but it may no longer be in the current code, as I didn't find it (yet).
Seems that most (or all?) I2C plugins add an Address selector themselves, as a limited set of addresses is usually supported.
I'll insert an extra header there.

> At least the _CPlugin code must then also perform the switching if we also need this one to operate behind the multiplexer.

Will add support for that, didn't see/know that I2C support was there too.

Sounds like fun 👍 just give me a couple of hours to fix all this. Thanks for the great feedback!

TD-er commented:
Maybe first leave the Cplugin part of the code, as that one doesn't have proper device defines like we have in the plugin parts.
N.B. the only controller we now have that actually does address hardware, is the TTN RN2483 LoRaWAN controller.
And I guess it may need a lot more changes while it isn't even needed for the I2C to UART bridge as that one supports 16 or 32 addresses.

About the need to reset the multiplexer.
The datasheet mentions when a bus gets locked.
This does happen every now and then, especially with the Wemos motor controller if it hasn't got the (really needed) firmware upgrade.

TD-er commented:
> > * All multiplexer ports disabled is like "direct connected", but the main I2C bus is always connected.
> 
> Don't exactly understand what you're trying to say here?
> 

There is the main I2C bus on the ESP.
This one has only the multiplexer in your setup, but it also will have all other I2C devices not connected behind the multiplexer.
Take the use case where you try to scan the I2C bus:
- Scan the main bus, without multiplexer channels enabled.
- Scan multiplexer bus 1 ... N

All devices you do see in the first step cannot be connected via the multiplexer.
Meaning if a specific address is found on the first scan, is only connected directly to the ESP and thus cannot be used behind a multiplexer.
So the only possibility to have I2C address duplicates is when that address is not seen on the first scan.
If it is seen on the first scan, then it will also be seen on every other multiplexer port.

TD-er commented:
I just tested a setup using 3 OLED displays on the same I2C address, running Framed OLED plugin.  (On an ESP8266)
Well the multiplexing works, but the Framed OLED plugin does take way too much RAM to be useful with 3 instances, so that's one thing we may need to look into. 
But that's outside scope of this PR.

tonhuisman committed:
[I2C] Improve I2C Multiplexer support

tonhuisman commented:
A quite major update on the I2C Multiplexer support (and some other I2C related improvements)
- General: Many bugs/issues reported, and some I found, are fixed
- I2C Scanner: Don't repeatedly report devices on main I2C bus for each channel
- Hardware page: Added setting for `Slow device Clock Speed`, defaults to 100000kHz (generic I2C)
- Hardware page: Added setting for `GPIO -> Reset:(optional)` pin of an I2C Multiplexer, with initialization and supporting function
- Devices list: Show multiple I2C multiplexer channels when configured
- Device Edit page: Add checkbox to `Force Slow I2C speed` for a single I2C device, and add extra header `Device Settings` below I2C options (generic I2C)
- Device Edit page: Allow Single channel/Multiple channels on I2C Multiplexer, reloads the page on change, and offers multiple checkboxes for Multiple channels setting

Some updated screenshots:
*Hardware page without I2C Multiplexer support:*
![Hardware_page_no_mux](https://user-images.githubusercontent.com/1922393/90315701-c3c90c00-df1d-11ea-8dc7-f6d40cda7b44.png)

*Device BME280 without Mux support:*
![Device_bme280_no_mux](https://user-images.githubusercontent.com/1922393/90315714-e6f3bb80-df1d-11ea-852f-e07c88e7d053.png)

*Hardware page with I2C Multiplexer support:*
![Hardware_page_with_mux_reset](https://user-images.githubusercontent.com/1922393/90315720-ef4bf680-df1d-11ea-8907-66cbaad6bd30.png)

*Devices list with multiple and single channel Mux configured:*
![Devices_list_mux_multiple_channels](https://user-images.githubusercontent.com/1922393/90315739-15719680-df1e-11ea-82e2-2b4a41dc64d8.png)

*Device edit page with Multiple channels selected:*
![Device_edit_mux_multiple_channels](https://user-images.githubusercontent.com/1922393/90315751-2fab7480-df1e-11ea-9eb0-375ff8e52cfb.png)

Tested lightly with the devices I have available, feedback is welcome 👍

tonhuisman committed:
Merge branch 'mega' into feature/I2C_multiplexer

tonhuisman committed:
[I2C] Fix builderror on SettingsStructSize (again)

TD-er commented on src/WebServer_I2C_Scanner.ino:
```diff
@@ -176,6 +239,77 @@ String getKnownI2Cdevice(byte address) {
+#endif
+      Wire.beginTransmission(address);
+      error = Wire.endTransmission();
+
```
Need a `delay(1);` here, as you would otherwise get Watchdog resets in an I2C scan.

TD-er replied:
The I2C scan now takes 14 seconds, which is way too long.
Not sure why, as it is way more than 9x the time it used to take.

TD-er replied:
Ah found it.. the default settings set the low clock speed to 100, instead of 100 kHz.
I think that does make a difference :)

tonhuisman replied:
I mostly tested on an ESP32, and didn't have issues there. Now I'm testing on ESP8266, and it is dead slow indeed. Think you've pinpointed the cause, below. Working on it.

tonhuisman replied:
> 
> 
> Ah found it.. the default settings set the low clock speed to 100, instead of 100 kHz.
> I think that does make a difference :)

That is because I didn't update the Build number again after adding this new setting, I had the same issue on the ESP32 but changed it immediately to 100000 kHz, but on the ESP8266 I forgot to update/check the value, and the minimum input value is 100, that's why it was at 100 😬 In a normal update from below build 20110 it should be set at 100000 kHz.

TD-er commented:
Can you change "Multiplexer SD1/SC1" to something like "Multiplexer port 1" ?
That's easier to read.

TD-er commented on src/WebServer_I2C_Scanner.ino:
```diff
@@ -5,72 +5,135 @@
+      , int8_t channel
+      , int nDevices
+#ifdef FEATURE_I2CMULTIPLEXER
+      , String &excludeDevices
```
Why use a String here?
A `std::vector<bool>` is a lot cheaper to use here.
A vector of bool is a special kind of vector, which uses a packed representation.
So of length 128 does only use 128 bits, which is 16 bytes (+ the few bytes the vector container uses)

Searching through a String of numbers separated with a `|` roughly 1000 times is quite inefficient.

tonhuisman replied:
Must be my lack on C++ knowledge, I'm reading up on this `<vector>` stuff.

TD-er replied:
Just the very basics:
- You have to set the length using `resize` yourself, or else you will get out of range issues when addressing it.
- Address the values simply by `variable[x]`.

For example:

```c++
typedef std::vector<bool> i2c_addresses_t;

i2c_addresses_t i2c_address;
i2c_address.resize(128);
for (int i = 0; i < 128; ++i) {
  i2c_address[i] = false;
}
```

You can also use `+=` to increment its size, but that's not very efficient due to resizes, and you may run in out of range addressing.
To make the `+=` more efficient, you can call `reserve`. Just like how it is done in the `String` class.

tonhuisman committed:
[I2C] Bugfixes and improvements for multiplexer implementation

tonhuisman commented:
Several bugfixes and improvements again

- Unexpected resets during I2C Scanner operation solved by adding `delay(1)` and changed as suggested from `String` with expensive `indexOf` operation to `stdd:vector<bool>`

> Can you change "Multiplexer SD1/SC1" to something like "Multiplexer port 1" ?
> That's easier to read.

- Renamed the settings variable from `I2C_Multiplexer_Port` to `...Channel`
- Chose term 'channel' over 'port' as a port feels like a single-connection endpoint, and a channel more like a multiple-connection endpoint

Updated screenshots will go in the documentation

TD-er commented:
You never resize the `excludeDevices`.
You do check for the size to be 0, before calling the `[]` operator, but you never add something to the vector.
It also doesn't make sense to check for the size, just resize it to the number of addresses 

```c++
int scanI2CbusForDevices( // Utility function for scanning the I2C bus for valid devices, with HTML table output
        int8_t muxAddr
      , int8_t channel
      , int nDevices
#ifdef FEATURE_I2CMULTIPLEXER
      , i2c_addresses_t &excludeDevices
#endif
) {
  byte error, address;
#ifdef FEATURE_I2CMULTIPLEXER
  bool skipCheck = false;
  if (muxAddr != -1) {
    excludeDevices.resize(128);
    for (int i = 0; i < 128; ++i) {
      excludeDevices[i] = false;
    }
  }
#endif

  for (address = 1; address <= 127; address++)
  {
[...]
```

tonhuisman commented:
> You never resize the `excludeDevices`.

The loop across all addresses will always go from 1 to 127, so when accessing the bool I need to have all 128 indexes/items, IMHO.

> You do check for the size to be 0, before calling the `[]` operator, but you never add something to the vector.

That is just a precaution, might it be called with an empty argument (quite unlikely, that's true).

> It also doesn't make sense to check for the size, just resize it to the number of addresses

Well, I either resize the vector to the last index value actually true, or do no resizing and simply access the indexed value. I chose simplicity, and smaller code size, over memory efficiency here, as the memory should be released after the action anyway. Or does that vector need some kind of destruction after use? In that case I'd resize it to 0 at the end.

TD-er commented:
Ah found my error... I only looked on the name used inside the function, not where it was called....
Sorry

To answer your other remark...
The check for the size of the vector can be done outside the loop.

And another remark, please allocate the variables as close to when they are needed.
This does reduce their scope and also reduces the chance for errors.
For example the check when to skip is declared at the begin of the function, so it is kept the last value, even when it isn't updated inside the if statement.
That's typical for some place where an error may occur.
Also the variables are not initialized when declared, which can also lead to errors later on as you may move some code around.

tonhuisman commented:
> To answer your other remark...
> The check for the size of the vector can be done outside the loop.

I still need to loop that loop, even if `size()` is 0. And that `size()` call is about as costly as introducing an extra `bool` that needs to be checked. IMHO, that won't make much, if any, difference.

TD-er commented:
I'm not entirely sure what the 'cost' is for a call to `size()`.
I can't find its implementation.
If it is `std::distance(begin(), end())`, then it is at least a subtraction of 2 pointers and a divide by the size of a single element.
If it is a member variable of the container, then it is hardly more expensive compared to a local bool.

tonhuisman committed:
[I2C] Optimizations and improvements

tonhuisman committed:
Merge branch 'mega' into feature/I2C_multiplexer

tonhuisman committed:
[I2C/Doc] Add Hardware page documentation, RTD help button

tonhuisman committed:
Merge branch 'mega' into feature/I2C_multiplexer

TD-er commented:
Wow, documentation 👍 

Have you seen the other PR I made for the BME280?
Could you also test that PR along with this PR?

tonhuisman commented:
> Wow, documentation 👍

Well, the RTD page for Hardware didn't exist, and the wiki docs didn't make much sense to me, so I started a fresh new page 😉 

> Have you seen the other PR I made for the BME280?
> Could you also test that PR along with this PR?

Yes, I saw that, but I'm fresh out of BME280 sensors, and though I already ordered a couple from Ali 2 weeks ago, they haven't arrived yet.
Also, my github fu isn't that strong yet, so I don't know how to pull that PR into my workspace (or to pull only that PR). Can you provide a git command-line for that, please (my command-line skills are quite ok 😉 )

TD-er commented:
Well I'm not a git guru myself, but I do manage a lot using GitKraken as my Git tool.
As I've said before, that's the tool that finally makes me understand Git although I've been using git for a few years professionally before I found GitKraken.

But I tested the changes here on my board, so I can merge it and you can update this branch with the mega branch changes.

TD-er committed:
Merge branch 'mega' into feature/I2C_multiplexer

tonhuisman committed:
[I2C] Some fixes and improvements

tonhuisman committed:
[I2C] Add TCA9543a multiplexer, updated docs

tonhuisman commented:
Added support for the TCA9543a 2 channel I2C multiplexer because the board is much smaller than the TCA9546/8 boards, digitally compatible (but with only 2 channels and 4 I2C addresses), and it *is* available from Ali and EBay, where the TCA9546a and PCA9540 are much harder to find. Chip-only is available for all of these chips, but you'd have to do much more/precise soldering to use these 😉.
And manufacturer documentation (google for the chip type + specs) shows that the TCA9544a and TCA9545a are quite compatible with the TCA9546a 4 channel I2C multiplexer, but the additional Interrupt feature is currently not supported, and TCA9544a doesn't have a Reset pin.

Anyone has suggestions/requests for other I2C multiplexers (though IMHO we now have enough options 😄) ?

TD-er commented:
> Anyone has suggestions/requests for other I2C multiplexers (though IMHO we now have enough options 😄) ?

I think we should merge it as soon as you think it is working now.
I did test some with the 8-pin version and that one is working fine right now.

Maybe I missed it, but there should be a remark in the documentation that the I2C address of the multiplexer should not conflict with any of the chips you intend to use. This in contrast with the chips used after the multiplexer, which may have conflicting addresses with other chips, as long as they are used "behind" the multiplexer.

tonhuisman commented:
> Maybe I missed it, but there should be a remark in the documentation that the I2C address of the multiplexer should not conflict with any of the chips you intend to use.

I'll check if I wrote that clear enough and update the docs if needed.

Oh, and the mailman just delivered the BME280's I ordered, so, some more stuff to ~~play~~ test with 😆

TD-er commented:
Please let me know how well that's working.

I do have a few sensors here that are not working well.
Both BMP280 and BME280... or at least that's what they should be.
In the logs you continously see them detecting the chip and the next second not being able to detect the chip ID.
As a result the reported pressure is quite low (since the reading procedure cannot be completed).

Those sensors known to be working fine still work OK and older builds show the same behavior on these sensors.
So let me know, if this is just a bad batch here, or maybe the reseller channels are getting flooded with these badly behaving ones.  (bad clones perhaps?)

tonhuisman committed:
[I2C Docs] Emphasize configuration with multiplexer

tonhuisman commented:
I've been testing for a short while now with 2 BME280s on different multiplexer channels, and all is working as expected.
I first tested them separately, without mux, and monitored the log for any errors, but all was working just fine.
Then I added them to my 'christmas tree' I2C test-bed (pretty hard to get all devices connected to power and SDA/SCL), and all is still working as expected, the only way I can induce those `unable to detect Chip ID` errors is when I select a channel where no BME is connected. Looks like you got some from a 'monday-morning' batch 😞 

Here's the 'christmas tree' I2C Scan (on an ESP32)
![Screenshot - 19-08-2020 , 17_34_51](https://user-images.githubusercontent.com/1922393/90655872-6648fe00-e242-11ea-845c-364fadd56862.png)

And a photo of that 'christmas tree':
![IMG_20200819_171054](https://user-images.githubusercontent.com/1922393/90654859-1f0e3d80-e241-11ea-9753-8f2135b1c3eb.jpg)


The readings on those BME sensors look quite alright to me (except that the temperature is a big high here, but it sounds reasonable 😬)
![Screenshot - 19-08-2020 , 17_45_14](https://user-images.githubusercontent.com/1922393/90658499-f0459680-e243-11ea-82e1-0c68d68a5739.png)

tonhuisman committed:
[I2C Docs] Add note about multiplexer address

tonhuisman committed:
Merge branch 'mega' into feature/I2C_multiplexer

tonhuisman commented:
All seems stable here, so I'll resume work on PR #3205.

TD-er committed:
[I2C] Add PLUGIN_WEBFORM_SHOW_I2C_PARAMS to separate I2C address setup

Also made an example for P036 Framed OLED.

TD-er committed:
Merge remote-tracking branch 'tonhuisman/feature/I2C_multiplexer' into feature/I2C_multiplexer

TD-er commented:
I did add an example for how to split the I2C address selection.
I will add more commits for all other I2C plugins.

TD-er committed:
[I2C] Move I2C address selector for all I2C plugins with address field

tonhuisman commented on src/_P095_ILI9341.ino:
```diff
@@ -164,7 +164,7 @@ boolean Plugin_095(byte function, struct EventStruct *event, String& string)
       {
         Device[++deviceCount].Number = PLUGIN_ID_095;
-        Device[deviceCount].Type = DEVICE_TYPE_I2C;
+        Device[deviceCount].Type = DEVICE_TYPE_I2C;    // FIXME TD-er: This is no I2C, but looks like SPI.
```
This is already covered by #3205

TD-er replied:
Great.
Then I will remove this from the PR.

TD-er committed:
[Cleanup] Uncrustify all I2C plugins

When adding the new webform I2C params function, it struck me how inconsistent the formatting of our code is.
So how looks like a good moment to uncrustify all I2C plugins as their code is touched anyway.

TD-er committed:
Revert changes in P095 as they are already already covered by #3205

TD-er committed:
[I2C multiplexer] Fix build issue on units not using I2C multiplexer

TD-er committed:
[Build] Fix build size exceeded for testing builds
